include/ginkgo/core/matrix/bccoo.hpp

```diff
@@ -0,0 +1,417 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2021, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_MATRIX_BCCOO_HPP_
+#define GKO_PUBLIC_CORE_MATRIX_BCCOO_HPP_
+
+
+#include <cstring>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+
+
+namespace gko {
+
+
+/**
+ * @brief The matrix namespace.
+ *
+ * @ingroup matrix
+ */
+namespace matrix {
+
+
+template <typename ValueType, typename IndexType>
+class Coo;
+
+
+template <typename ValueType, typename IndexType>
+class Csr;
+
+
+template <typename ValueType>
+class Dense;
+
+/* // JIAE
+template <typename ValueType, typename IndexType>
+class BccooBuilder;
+*/
+
+/**
+ * BCCOO is a matrix format which only stores the nonzeros coeffficients
+ * by blocks of consecutive elements.
+ *
+ * First the elements are sorted by row and column indexes, and, then,
+ * only the pairs (column,value) is stored in the 1D array of bytes.
+ * The column indexes could be stored as their value or as the difference
+ * from the previous element in the same row.
```

Comment:
This is not clear to me, are both options implemented, then it should read as
```suggestion
 * The column indexes can be stored directly or as the difference
 * from the previous element in the same row.
```

Reply:
Done